File: sdks/python/boxlite/sync_api/_simplebox.py
"""
SyncSimpleBox - Synchronous wrapper for SimpleBox.

Provides a synchronous API for box operations.
API mirrors async SimpleBox exactly.
"""

from typing import TYPE_CHECKING, Dict, Optional

from ..exec import ExecResult

if TYPE_CHECKING:
    from ._boxlite import SyncBoxlite
    from ._box import SyncBox

__all__ = ["SyncSimpleBox"]


class SyncSimpleBox:
    """
    Synchronous wrapper for SimpleBox.

    Provides synchronous methods for executing commands in a BoxLite container.
    Uses SyncBox internally which handles async bridging via greenlet.
    API mirrors async SimpleBox exactly.

    Usage (standalone - recommended):
        with SyncSimpleBox(image="python:slim") as box:
            result = box.exec("ls", "-la")
            print(result.stdout)

    Usage (with explicit runtime):
        with SyncBoxlite.default() as runtime:
            with SyncSimpleBox(image="python:slim", runtime=runtime) as box:
                result = box.exec("ls", "-la")
                print(result.stdout)
    """

    def __init__(
        self,
        image: str,
        memory_mib: Optional[int] = None,
        cpus: Optional[int] = None,
        runtime: Optional["SyncBoxlite"] = None,
        name: Optional[str] = None,
        auto_remove: bool = True,
        reuse_existing: bool = False,
        **kwargs,
    ):
        """
        Create a SyncSimpleBox.

        Args:
            image: Container image to use (e.g., "python:slim", "ubuntu:latest")
            memory_mib: Memory limit in MiB (default: system default)
            cpus: Number of CPU cores (default: system default)
            runtime: Optional SyncBoxlite runtime. If None, creates default runtime.
            name: Optional unique name for the box
            auto_remove: Remove box when stopped (default: True)
            reuse_existing: If True and a box with the given name already exists,
                reuse it instead of raising an error (default: False)
            **kwargs: Additional BoxOptions parameters
        """
        from ._boxlite import SyncBoxlite
        from ..boxlite import BoxOptions

        # Handle optional runtime
        if runtime is None:
            runtime = SyncBoxlite.default()
            self._owns_runtime = True
        else:
            self._owns_runtime = False

        self._runtime = runtime

        # Create box options
        self._box_opts = BoxOptions(
            image=image,
            cpus=cpus,
            memory_mib=memory_mib,
            auto_remove=auto_remove,
            **kwargs,
        )

        # Store for lazy creation in __enter__
        self._name = name
        self._reuse_existing = reuse_existing
        self._box: Optional["SyncBox"] = None
        self._created: Optional[bool] = None

    def __enter__(self) -> "SyncSimpleBox":
        """Enter context - starts runtime if owned, then starts the box."""
        # Start runtime if we own it
        if self._owns_runtime:
            self._runtime.start()

        # Create or reuse box via runtime
        if self._reuse_existing:
            self._box, self._created = self._runtime.get_or_create(
                self._box_opts, name=self._name
            )
        else:
            self._box = self._runtime.create(self._box_opts, name=self._name)
            self._created = True
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        """Exit context - stops the box, then stops runtime if owned."""
        # Stop the box (SyncBox.stop() is already sync)
        if self._box is not None:
            self._box.stop()

        # Stop runtime if we own it
        if self._owns_runtime:
            self._runtime.stop()

    @property
    def id(self) -> str:
        """Get the box ID."""
        return self._box.id

    @property
    def name(self) -> Optional[str]:
        """Get the box name (if set)."""
        return self._box.name

    @property
    def created(self) -> Optional[bool]:
        """Whether this box was newly created (True) or an existing box was reused (False).

        Returns None if the box hasn't been started yet.
        """
        return self._created

    def info(self):
        """Get box information."""
        return self._box.info()

    def exec(
        self,
        cmd: str,
        *args: str,
        env: Optional[Dict[str, str]] = None,
    ) -> ExecResult:
        """
        Execute a command in the box synchronously.

        Args:
            cmd: Command to run (e.g., "ls", "python")
            *args: Command arguments (e.g., "-l", "-a")
            env: Environment variables as dict

        Returns:
            ExecResult with exit_code, stdout, and stderr

        Example:
            result = box.exec("ls", "-la")
            print(f"Exit code: {result.exit_code}")
            print(f"Output: {result.stdout}")
        """
        # Convert args to list format expected by SyncBox
        arg_list = list(args) if args else None
        env_list = list(env.items()) if env else None

        # SyncBox.exec() returns SyncExecution - already sync!
        execution = self._box.exec(cmd, arg_list, env_list)

        # Collect stdout (sync iteration)
        stdout_lines = []
        for line in execution.stdout():
            if isinstance(line, bytes):
                stdout_lines.append(line.decode("utf-8", errors="replace"))
            else:
                stdout_lines.append(line)

        # Collect stderr (sync iteration)
        stderr_lines = []
        for line in execution.stderr():
            if isinstance(line, bytes):
                stderr_lines.append(line.decode("utf-8", errors="replace"))
            else:
                stderr_lines.append(line)

        # Wait for completion (sync)
        result = execution.wait()

        return ExecResult(
            exit_code=result.exit_code,
            stdout="".join(stdout_lines),
            stderr="".join(stderr_lines),
            error_message=result.error_message,
        )

    def stop(self) -> None:
        """Stop the box (preserves state for restart)."""
        self._box.stop()

    def metrics(self):
        """Get box metrics (CPU, memory usage)."""
        return self._box.metrics()

File: examples/node/simplebox.js
/**
 * SimpleBox Example - Basic command execution
 *
 * Demonstrates:
 * - Creating a simple container
 * - Executing commands
 * - Collecting output
 * - Proper cleanup
 */

import { SimpleBox } from '@boxlite-ai/boxlite';

async function main() {
  console.log('=== SimpleBox Example ===\n');

  // Create a box with Alpine Linux
  const box = new SimpleBox({ image: 'alpine:latest' });

  try {
    console.log('1. Running basic command...');
    const result1 = await box.exec('echo', 'Hello from BoxLite!');
    console.log(`   Output: ${result1.stdout.trim()}`);
    console.log(`   Exit code: ${result1.exitCode}\n`);

    console.log('2. Listing files...');
    const result2 = await box.exec('ls', '-la', '/');
    console.log(`   Output (first 5 lines):`);
    const lines = result2.stdout.split('\n').slice(0, 5);
    lines.forEach(line => console.log(`   ${line}`));
    console.log();

    console.log('3. Running command with environment variables...');
    const result3 = await box.exec('sh', '-c', 'echo $MY_VAR', { MY_VAR: 'custom-value' });
    console.log(`   Output: ${result3.stdout.trim()}\n`);

    console.log('4. Box metadata...');
    const info = box.info();
    console.log(`   ID: ${box.id}`);
    console.log(`   Name: ${box.name || '(unnamed)'}`);
    console.log(`   Info: ${JSON.stringify(info, null, 2)}\n`);

    console.log('✅ All commands completed successfully!');

    console.log('\n5. Reuse existing box by name...');
    const box2 = new SimpleBox({ image: 'alpine:latest', name: 'reuse-demo', reuseExisting: true });
    try {
      const r1 = await box2.exec('echo', 'first open');
      console.log(`   Created: ${box2.created}`);
      console.log(`   Output: ${r1.stdout.trim()}`);

      // Second open reuses the same box
      const box3 = new SimpleBox({ image: 'alpine:latest', name: 'reuse-demo', reuseExisting: true });
      try {
        const r2 = await box3.exec('echo', 'second open');
        console.log(`   Reused (created=${box3.created}): ${r2.stdout.trim()}`);
      } finally {
        await box3.stop();
      }
    } finally {
      await box2.stop();
    }
  } finally {
    console.log('\n6. Cleaning up...');
    await box.stop();
    console.log('   Box stopped and removed.');
  }
}

// Run the example
main().catch(error => {
  console.error('Error:', error);
  process.exit(1);
});
